### PR TITLE
Filling the ndarray

### PR DESCRIPTION
**Pull request by arushanova**
Filling the ndarray with reconstructed energy and position from the root file. The time dependence is done via adding weights to the events.

### CODE REVIEW

File: echidna/core/fill_spectrum.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      initial_time (int): The beginning of time period to study
+      final_time (int): The end time of time period to study
+      T (float): The Half-life of a studied background
+
+    Returns:
+      (*tuple*). Times (*list* of *int*) and Weights (*list* of *float*)
+    """
+    times = [0]
+    weights = [0]
+    for time_step in range(initial_time, final_time):
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: I think the step should be `Spectra._time_width` and the initial and final times should always be `Spectra._time_low` and `Spectra._time_high`. Or better still something like:
```python
for time_step in range(0, Spectra._time_bins):
    time = time_step * Spectra._time_width + Spectra._time_low
```

File: echidna/core/fill_spectrum.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    """This function fills in the ndarray of energies, radii, times 
+    and weights. It takes the reconstructed energies and positions
+    of the events from the root file. In order to keep the statistics, 
+    the time dependence is performed via adding weights to every event
+    depending on the time period. Both, studied time and Half-life must 
+    be written in the same units.  
+    
+    Args:
+      filename (str): A root file to study 
+      spectrum (spectra.Spectra): Ndarray to be filled
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: I think this function should take a `name` (or construct the name from the filename) and return a constructed Spectra.

File: echidna/core/fill_spectrum.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    print spectrumname
+    spectrum = spectra.Spectra(str(spectrumname))
+
+    times = [0]
+    for time_step in range(0, spectrum._time_bins):
+        time = time_step * spectrum._time_width + spectrum._time_low 
+        times.append(time)
+
+    if 'AV' in spectrumname:
+        print "AV WEIGHTS ARE CURRENTLY UNAVAILABLE"
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: Probably no longer true...

File: echidna/core/fill_spectrum.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    depending on the time period. Both, studied time and Half-life must 
+    be written in the same units.  
+    
+    Args:
+      filename (str): A root file to study 
+      spectrumname (str): A name of future ndarray 
+      T (float): The Half-life of a studied background
+
+      Returns:
+        spectrum (spectra.Spectra)
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: If you format this as 

    Returns:
      spectrum (:class:`echidna.core.spectra.Spectra`)

sphinx should be able to understand the type.

**Comment by pgjones** [comment]
Would be keen to merge, although my request #11 needs to be merged first.

File: echidna/core/fill_spectrum.py (hunk 3 above)
Comment (arushanova) [inline echidna/core/fill_spectrum.py]: All AV weights are equal to 1, so AV weighting is not applied properly.

File: echidna/core/fill_spectrum.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+    for ds, run in rat.dsreader(filename):
+        if ds.GetEVCount() == 0:
+            continue
+        ev = ds.GetEV(0)       
+        vertex = ev.GetFitResult("scintFitter").GetVertex(0)
+        if not vertex.ContainsEnergy() or not vertex.ValidEnergy():
+            continue
+        if not vertex.ContainsPosition() or not vertex.ValidPosition():
+            continue
+        if vertex.GetPosition().Mag() > 3500.0:
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: I don't think this condition should exist.

File: echidna/core/fill_spectrum.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        for iparticle in range(0, particle_count):
+            the_particle = mc.GetMCParticle(iparticle)
+            magnitude = the_particle.GetPosition().Mag()
+
+        vertex = ev.GetFitResult("scintFitter").GetVertex(0)
+        if not vertex.ContainsEnergy() or not vertex.ValidEnergy():
+            continue
+        if not vertex.ContainsPosition() or not vertex.ValidPosition():
+            continue
+        if vertex.GetPosition().Mag() > 3500.0:
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: Dito

File: echidna/core/fill_spectrum.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+            continue
+
+        for time,weight in zip(times, weights):
+            spectrum.fill(mc.GetScintEnergyDeposit(), magnitude, time, weight)
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: I think the quenched energy deposited makes more sense.

File: echidna/core/fill_spectrum.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        if ds.GetEVCount() == 0:
+            continue
+        ev = ds.GetEV(0) 
+        mc = ds.GetMC()
+        particle_count = mc.GetMCParticleCount()
+        for iparticle in range(0, particle_count):
+            the_particle = mc.GetMCParticle(iparticle)
+            magnitude = the_particle.GetPosition().Mag()
+
+        vertex = ev.GetFitResult("scintFitter").GetVertex(0)
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: I think you should remove all references and use of the vertex

**Comment by pgjones** [comment]
Thanks for fixing my :bug:

File: echidna/core/fill_spectrum.py [hunk 9]
```diff
@@ -0,0 +1,132 @@
+import ROOT
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: I would remove this, as not needed

File: echidna/core/fill_spectrum.py [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+    else:
+        weights = _scint_weights(times, T)
+
+    for ds, run in rat.dsreader(filename):
+        if ds.GetEVCount() == 0:
+            continue
+        ev = ds.GetEV(0) 
+        mc = ds.GetMC()
+        particle_count = mc.GetMCParticleCount()
+        for iparticle in range(0, particle_count):
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: I suggest instead, 
```
    magnitude = mc.GetMCParticle(0).GetPosition().Mag()
````
@jwaterfield, @ashleyrback consensus needed.

File: echidna/core/fill_spectrum.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+    if 'AV' in spectrumname:
+        print "AV WEIGHTS ARE CURRENTLY UNAVAILABLE"
+        weights = _av_weights(times, T)
+    else:
+        weights = _scint_weights(times, T)
+
+    for ds, run in rat.dsreader(filename):
+        if ds.GetEVCount() == 0:
+            continue
+        ev = ds.GetEV(0)
```
Comment (pgjones) [inline echidna/core/fill_spectrum.py]: I think you should loop over the EVs, @jwaterfield, @ashleyrback consensus needed.

**Pull request merged**